### PR TITLE
[serdes] extract-nested instead of relying on hydration

### PR DESCRIPTION
**Pull request by sansolovyov**
Serialization mostly relies on hydration to get access to nested data, and it's not very consistent: sometimes hydration is done for the convenience of developer and contains surprising data that's hard to export in a generalized way.

This PR replaces that with a new behavior, which traverses `:transform` of a serialization spec for the model and fetches required nested data recursively. Also, it automatically detects if a model supports being filtered by collection.

References #43068

It makes sense to read commits separately since first one is mostly "pre-factoring".

### CODE REVIEW

File: enterprise/backend/test/metabase_enterprise/serialization/v2/extract_test.clj [hunk 1]
```diff
@@ -1706,3 +1706,44 @@
...
+        (is (=? (assoc dc1 :series [s])
+                (u/rfirst (serdes/extract-query "DashboardCard" {:where [:= :id (:id dc1)]})))))
+      (let [spec (serdes/make-spec "Dashboard" nil)]
+        (is (= {(:id d) [(assoc dc1 :series [s])]}
+               (#'serdes/transform->nested (-> spec :transform :dashcards) [d])))
+        (is (=? (assoc d
+                       :dashcards [(assoc dc1 :series [s])]
+                       :tabs nil)
+                (u/rfirst (serdes/extract-query "Dashboard" {:where [:= :id (:id d)]})))))))
+  (testing "extract-nested is efficient"
```
Comment (crisptrutski) [inline enterprise/backend/test/metabase_enterprise/serialization/v2/extract_test.clj:1729]: nit: given this test is totally independent, might as well put it in a separate deftest

**Review by crisptrutski — APPROVED** [review]
I find this so much simpler now, the efficiency is just the icing ❤️ 🧁 

My one concern is just "kicking the tires" for the partitioning. Do we have an e2e test pushing enough data through here? Should we do a manual test otherwise?

File: enterprise/backend/test/metabase_enterprise/serialization/v2/extract_test.clj [hunk 2]
```diff
@@ -1706,3 +1706,65 @@
...
+(deftest extract-nested-test
+  (testing "extract-nested working"
+    (mt/with-temp [:model/Dashboard           d   {:name "Top Dash"}
+                   :model/Card                c1  {:name "Some Card"}
+                   :model/Card                c2  {:name "Some Inner Card"}
+                   :model/DashboardCard       dc1 {:dashboard_id (:id d) :card_id (:id c1)}
+                   :model/DashboardCardSeries s   {:dashboardcard_id (:id dc1) :card_id (:id c2)}]
+      (let [spec (serdes/make-spec "DashboardCard" nil)]
+        (is (= {(:id dc1) [s]}
+               (#'serdes/transform->nested (-> spec :transform :series) [dc1])))
```
Comment (camsaul) [inline enterprise/backend/test/metabase_enterprise/serialization/v2/extract_test.clj:1719]: Didn't you update `transform->nested` to take three args now? Not sure how this test passed

Comment (piranha) [reply enterprise/backend/test/metabase_enterprise/serialization/v2/extract_test.clj:1719]: oh wow, wtf, how did it pass indeed? 🤦‍♂️

**Pull request merged**